replay/models/kl_ucb.py

```diff
@@ -0,0 +1,156 @@
+from scipy.optimize import root_scalar
+
+
+class KLUCB(UCB):
+    """
+    Bernoulli `bandit model
+    <https://en.wikipedia.org/wiki/Multi-armed_bandit>`_. Same to :class:`UCB`
+    computes item relevance as an upper confidence bound of true fraction of
+    positive interactions.
+
+    In a nutshell, KL-UCB сonsiders the data as the history of interactions
+    with items. The interaction may be either positive or negative. For each
+    item the model computes empirical frequency of positive interactions
+    and estimates the true frequency with an upper confidence bound. The higher
+    the bound for an item is the more relevant it is presumed.
+
+    The upper bound below is what differs from the classical UCB. It is
+    computed according to the `original article
+    <https://arxiv.org/pdf/1102.2490.pdf>`_ where is proven to produce
+    assymptotically better results.
+
+    .. math::
+        u_i = \\max q \\in [0,1] :
+        n_i \\cdot \\operatorname{KL}\\left(\\frac{p_i}{n_i}, q \\right)
+        \\leqslant \\log(n) + c \\log(\\log(n)),
+
+    where
+
+    :math:`u_i` -- upper bound for item :math:`i`,
+
+    :math:`c` -- exploration coeficient,
+
+    :math:`n` -- number of interactions in log,
+
+    :math:`n_i` -- number of interactions with item :math:`i`,
+
+    :math:`p_i` -- number of positive interactions with item :math:`i`,
+
+    and
+
+    .. math::
+        \\operatorname{KL}(p, q)
+        = p \\log\\frac{p}{q} + (1-p)\\log\\frac{1-p}{1-q}
+
+    is the KL-divergence of Bernoulli distribution with parameter :math:`p`
+    from Bernoulli distribution with parameter :math:`q`.
+
+    Being a bit trickier though, the bound shares with UCB the same
+    exploration-exploitation tradeoff dilemma. You may increase the `c`
+    coefficient in order to shift the tradeoff towards exploration or decrease
+    it to set the model to be more sceptical of items with small volume of
+    collected statistics. The authors of the `article
+    <https://arxiv.org/pdf/1102.2490.pdf>`_ though claim `c = 0` to be of the
+    best choice in practice.
+
+
+    As any other RePlay model, KL-UCB takes a log to fit on as a ``DataFrame``
+    with columns ``[user_idx, item_idx, timestamp, relevance]``. Following the
+    procedure above, KL-UCB would see each row as a record of an interaction
+    with ``item_idx`` with positive (relevance = 1) or negative (relevance = 0)
+    outcome. ``user_idx`` and ``timestamp`` are ignored i.e. the model treats
+    log as non-personalized - item scores are same for all users.
+
+    If ``relevance`` column is not of 0/1 initially, then you have to decide
+    what kind of relevance has to be considered as positive and convert
+    ``relevance`` to binary format during preprocessing.
+
+    To provide a prediction, KL-UCB would sample a set of recommended items for
+    each user with probabilites proportional to obtained relevances.
+
+    >>> import pandas as pd
```

Comment:
Please check the example, now all models should support Dataset, please see doc example here https://github.com/sb-ai-lab/RePlay/blob/main/replay/models/ucb.py